### PR TITLE
replication: Cancel correct request type when append fails.

### LINKED ISSUE #454: src/replication.c:457: getRequest: Assertion `req->type == type' failed.
Failed jepsen run: https://github.com/canonical/jepsen.dqlite/actions/runs/5500511402/jobs/10023467098

[jepsen-data-append-pause,disk-failure.zip](https://github.com/canonical/raft/files/11999567/jepsen-data-append-pause.disk-failure.zip)

### PR DESCRIPTION
**Pull request by MathieuBordere**
Fixes #454

### CODE REVIEW

File: src/replication.c [hunk 1]
```diff
@@ -476,19 +479,44 @@ static void appendLeaderCb(struct raft_io_append *req, int status)
...
-        struct raft_apply *apply;
         ErrMsgTransfer(r->io->errmsg, r->errmsg, "io");
-        apply =
-            (struct raft_apply *)getRequest(r, request->index, RAFT_COMMAND);
-        if (apply != NULL) {
-            if (apply->cb != NULL) {
-                apply->cb(apply, status, NULL);
+        for (unsigned i = 0; i < request->n; i++) {
+            const struct request *req = getRequest(r, request->index + i, -1);
+            if (!req) {
```
Comment (MathieuBordere) [inline src/replication.c:491]: not sure if we should assert `req` here, but the original code also ignored `NULL` requests.

File: src/replication.c [hunk 2]
```diff
@@ -476,19 +479,44 @@ static void appendLeaderCb(struct raft_io_append *req, int status)
...
+                continue;
+            }
+            switch (req->type) {
+                case RAFT_COMMAND: {
+                    struct raft_apply *apply = (struct raft_apply *)req;
+                    if (apply->cb) {
+                        apply->cb(apply, status, NULL);
```
[the remark this answers is not in the document]
Comment (MathieuBordere) [reply src/replication.c:498]: Should be easy enough to hit, let me give it a shot.

File: test/integration/test_replication.c [hunk 3]
```diff
@@ -1179,3 +1179,24 @@ TEST(replication, lastStoredLaggingBehindCommitIndex, setUp, tearDown, 0, NULL)
...
+
+    /* Server 0 will fail to persist entry 2, a barrier */
+    CLUSTER_IO_FAULT(0, 10, 1);
+
+    /* Server 0 gets elected and creates a barrier entry at index 2 */
+    CLUSTER_BOOTSTRAP;
+    CLUSTER_START;
+    CLUSTER_START_ELECT(0);
+
+    /* Cluster recovers. */
```
[the remark this answers is not in the document]
Comment (MathieuBordere) [reply test/integration/test_replication.c:1198]: It recovers from a recently elected leader that has a disk error, so it has stepped down and a new, stable, leader is elected.

File: include/raft/fixture.h [hunk 4]
```diff
@@ -138,16 +138,22 @@ RAFT_API unsigned raft_fixture_leader_index(struct raft_fixture *f);
...
  * be a voting one. Also, the @i'th server must be connected to a majority of
  * voting servers.
  */
+RAFT_API void raft_fixture_start_elect(struct raft_fixture *f, unsigned i);
+
+/**
+ * Calls raft_fixture_start_elect, but waits and asserts that the @i'th server
+ * has become the leader.
+ */
 RAFT_API void raft_fixture_elect(struct raft_fixture *f, unsigned i);
```
[the remark this answers is not in the document]
Comment (freeekanayaka) [reply include/raft/fixture.h:157]: The original idea around `fixture.h` was to have test helpers that could be used both by raft itself for its own testing and by consuming applications (dqlite) to put in place their tests.

FWIW I think this didn't turn out that well, and in a perspective v1.0 release of raft, it'd be better to remove `fixture.h` from the public API (there could be a deprecation period), and let application-level test helpers to evolve independently.

**Pull request merged**